tc/core/polyhedral/cuda/memory_promotion_heuristic.cc

```diff
@@ -666,60 +676,92 @@ void promoteGreedilyAtDepth(
+              scope,
+              partialSched,
+              threadSchedule)) {
+        continue;
+      }
+      // Check reuse within threads.
+      auto schedule = partialSched.flat_range_product(threadSchedule);
+      if (!hasReuseWithin(*group, schedule)) {
+        continue;
+      }
+
+      // TODO: if something is already in shared, but reuse it within one
+      // thread only, there is no point in keeping it in shared _if_ it
+      // gets promoted into a register.
+
+      scop.promoteGroup(
+          Scop::PromotedDecl::Kind::Register,
+          tensorId,
+          std::move(group),
+          scope,
+          partialSchedUM);
+    }
+  }
+}
+
 // Promote at the positions of the thread specific markers.
 void promoteToRegistersBelowThreads(MappedScop& mscop, size_t nRegisters) {
   using namespace tc::polyhedral::detail;
 
-  auto& scop = mscop.scop();
-  auto root = scop.scheduleRoot();
+  auto root = mscop.scop().scheduleRoot();
+  auto markers = findThreadSpecificMarkers(root);
 
-  {
-    auto markers = findThreadSpecificMarkers(root);
-
-    for (auto marker : markers) {
-      auto partialSched = prefixSchedule(root, marker);
-      // Pure affine schedule without (mapping) filters.
-      auto mapping = findThreadMappingAncestor(root, marker);
-      auto prefixSchedMupa = prefixScheduleMupa(root, mapping);
-      auto mapSchedMupa = infixScheduleMupa(root, mapping, marker);
-      auto partialSchedMupa = prefixSchedMupa.flat_range_product(mapSchedMupa);
-
-      // Because this function is called below the thread mapping marker,
-      // partialSched has been intersected with both the block and the thread
-      // mapping filters.   Therefore, groups will be computed relative to
-      // blocks and threads.
-      auto groupMap = TensorReferenceGroup::accessedWithin(
-          partialSched, scop.reads, scop.writes);
-      for (auto& tensorGroups : groupMap) {
-        auto tensorId = tensorGroups.first;
-
-        // TODO: sorting of groups and counting the number of promoted elements
-
-        for (auto& group : tensorGroups.second) {
-          auto sizes = group->approximationSizes();
-          // No point in promoting a scalar that will go to a register anyway.
-          if (sizes.size() == 0) {
-            continue;
-          }
-          if (!isPromotableToRegisterBelowThreads(
-                  *group, root, marker, prefixSchedMupa, mapSchedMupa)) {
-            continue;
-          }
-          if (!hasReuseWithin(*group, partialSchedMupa)) {
-            continue;
-          }
-          // TODO: if something is already in shared, but reuse it within one
-          // thread only, there is no point in keeping it in shared _if_ it
-          // gets promoted into a register.
-          scop.promoteGroup(
-              Scop::PromotedDecl::Kind::Register,
-              tensorId,
-              std::move(group),
-              marker,
-              partialSched);
-        }
-      }
-    }
```

Comment:
Can you do the outlining in a separate commit?
It's difficult to see exactly what you changed when you combine the move and the changes in a single commit.

Reply:
That would be very annoying because this would be another artificial commit with potential problems and potentially new code that will be erased by the subsequent commit.

Reply:
> That would be very annoying because this would be another artificial commit with potenti
al problems and potentially new code that will be erased by the subsequent commit.

Not if you do it properly.
(So not the way I initially pushed out.)